packages/app-mobile/utils/fs-driver/runOnDeviceTests.ts

```diff
@@ -0,0 +1,134 @@
+import Setting from '@joplin/lib/models/Setting';
+import shim from '@joplin/lib/shim';
+import uuid from '@joplin/lib/uuid';
+import { join } from 'path';
+import FsDriverBase from '@joplin/lib/fs-driver-base';
+import Logger from '@joplin/utils/Logger';
+
+const logger = Logger.create('fs-driver-tests');
+
+const expectToBe = async <T> (actual: T, expected: T) => {
+	if (actual !== expected) {
+		throw new Error(`Integration test failure: ${actual} was expected to be ${expected}`);
+	}
+};
+
+const testExpect = async () => {
+	// Verify that expect is working
+	await expectToBe(1, 1);
+	await expectToBe(true, true);
+};
+
+const testReadWriteFileUtf8 = async (tempDir: string) => {
+	logger.info('Testing fsDriver.writeFile and fsDriver.readFile with utf-8...');
+
+	const filePath = join(tempDir, uuid.createNano());
+
+	const testStrings = [
+		// ASCII
+		'test',
+
+		// Special characters
+		'𝐴 𝒕𝐞𝑺𝒕',
+
+		// Emojis
+		'✅ Test. 🕳️',
+	];
+
+	const testEncodings = ['utf-8', 'utf8', 'UtF-8'];
+
+	// Use the same file for all tests to test overwriting
+	for (const encoding of testEncodings) {
+		for (const testString of testStrings) {
+			const fsDriver: FsDriverBase = shim.fsDriver();
+			await fsDriver.writeFile(filePath, testString, encoding);
+
+			const fileData = await fsDriver.readFile(filePath, encoding);
+			await expectToBe(fileData, testString);
+		}
+	}
+};
+
+const testReadFileChunkUtf8 = async (tempDir: string) => {
+	logger.info('Testing fsDriver.readFileChunk...');
+
+	const filePath = join(tempDir, `${uuid.createNano()}.txt`);
+
+	const fsDriver: FsDriverBase = shim.fsDriver();
+
+	// 🕳️ is 7 bytes when utf-8 encoded
+	// à,á,â, and ã are each 2 bytes
+	const expectedFileContent = '01234567\nàáâã\n🕳️🕳️🕳️\ntest';
+	await fsDriver.writeFile(filePath, expectedFileContent, 'utf8');
+
+	const testEncodings = ['utf-8', 'utf8', 'UtF-8'];
+
+	for (const encoding of testEncodings) {
+		const handle = await fsDriver.open(filePath, 'r');
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 8, encoding), '01234567',
+		);
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 1, encoding), '\n',
+		);
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 8, encoding), 'àáâã',
+		);
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 8, encoding), '\n🕳️',
+		);
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 15, encoding), '🕳️🕳️\n',
+		);
+
+		// A 0 length should return null and not advance
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 0, encoding), null,
+		);
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 100, encoding), 'test',
+		);
+
+		// Should not be able to read past the end
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 10, encoding), null,
+		);
+
+		await expectToBe(
+			await fsDriver.readFileChunk(handle, 1, encoding), null,
+		);
+
+		await fsDriver.close(filePath);
+	}
+};
+
```

Comment:
**To-do**: Add a test for `fsDriver.appendFile`

Reply:
There's no definition for `appendFile` in [`FsDriverBase`](https://github.com/laurent22/joplin/blob/dev/packages/lib/fs-driver-base.ts). However, it does exist in [`fs-driver-node`](https://github.com/laurent22/joplin/blob/2fc7bcec0633c7e2c666a2a44f3857be3618d201/packages/lib/fs-driver-node.ts#L17) and we do [use it as a method on `fsDriver()`](https://github.com/laurent22/joplin/blob/2fc7bcec0633c7e2c666a2a44f3857be3618d201/packages/lib/services/e2ee/EncryptionService.ts#L523).

Additionally, `appendFile` in `fs-driver-rn` requires that the file already exists (at least for non-scoped storage, untested with scoped storage), while `appendFile` in `fs-driver-node` does not.

I've added a method definition to `FsDriverBase` and added a `TODO` comment for this `appendFile` behavior in `fs-driver-rn`.